gunrock/app/geo/geo_enactor.cuh

```diff
@@ -0,0 +1,490 @@
+            enactor_slices[this -> gpu_num * this -> enactor -> num_gpus + peer_];
+        
+        auto &enactor_stats    = enactor_slice.enactor_stats;
+        auto &graph            = data_slice.sub_graph[0];
+        auto &frontier         = enactor_slice.frontier;
+        auto &oprtr_parameters = enactor_slice.oprtr_parameters;
+        auto &retval           = enactor_stats.retval;
+        auto &iteration        = enactor_stats.iteration;
+
+        auto &locations_lat    = data_slice.locations_lat;
+        auto &locations_lon    = data_slice.locations_lon;
+
+        auto &latitude	       = data_slice.latitude;
+	auto &longitude	       = data_slice.longitude;
+
+	auto &valid_locations  = data_slice.valid_locations;
+	auto &active	       = data_slice.active;
+	auto &geo_iter	       = data_slice.geo_iter;
+
+	auto &geo_complete     = data_slice.geo_complete;
+
+	auto &D		       = data_slice.D;
+	auto &Dinv	       = data_slice.Dinv;
+	auto &W		       = data_slice.W;
+
+	util::Location target = util::DEVICE;
+ 
+        // --
+        // Define operations
+
+	// Custom spatial center kernel for geolocation
+
+	// printf("Gather operator ... \n");
+	// compute operation, substitute for neighbor reduce,
+	// ForAll() with a for loop inside (nested forloop),
+	// visiting all the neighbors and calculating the spatial center
+	// after a gather.
+	auto gather_op = [
+	    graph,
+	    locations_lat,
+	    locations_lon,
+	    latitude,
+	    longitude,
+	    valid_locations,
+	    iteration
+	] __host__ __device__ (VertexT *v_q, const SizeT &pos) {
+
+	    VertexT v 		= v_q[pos];
+	    SizeT start_edge 	= graph.CsrT::GetNeighborListOffset(v);
+	    SizeT num_neighbors = graph.CsrT::GetNeighborListLength(v);
+
+	    SizeT i 		= 0;
+
+	    for (SizeT e = start_edge; e < start_edge + num_neighbors; e++) {
+		VertexT u = graph.CsrT::GetEdgeDest(e);
+		if (util::isValid(latitude[u]) && util::isValid(longitude[u])) {
+		    // gather locations from neighbors	
+		    locations_lat[(v * num_neighbors) + i] = latitude[u];
+                    locations_lon[(v * num_neighbors) + i] = longitude[u];
+		    i++;
+		}
+	    }
+	    valid_locations[v] = i;
+
+	};
+
+	// printf("Spatial Median operator ... \n");
+	auto compute_op =  [
+	    graph,
+	    locations_lat,
+	    locations_lon,
+	    latitude,
+	    longitude,
+	    valid_locations,
+            iteration,
+	    active,
+	    D, Dinv, W
+	] __host__ __device__ (VertexT *v_q, const SizeT &pos) {
+
+	    VertexT v 		= v_q[pos];
+	    SizeT n		= graph.CsrT::GetNeighborListLength(v);
+
+	    // if no predicted location, and neighbor locations exists
+	    // Custom spatial center kernel for geolocation
+	    if (!util::isValid(latitude[v]) && !util::isValid(longitude[v])) {
+
+		spatial_center (locations_lat,
+				locations_lon,
+				n,
+				valid_locations[v],
+				latitude,
+				longitude,			
+				v,
+				D, Dinv, W,
+				false);
+
+	    } // -- median calculation.
+	};
+
+	auto status_op = [
```

Comment:
I think this could be done inside `gather_op` or `compute_op`, maybe saving an extra pass over the frontier

Reply:
I don't use this operator anyways, it is just if someone wants to do a complete geo run. Not used for performance measurement. Keeping it separate so it looks clean.